### PR TITLE
Add cuda version of SpatialBNOp also optimize SpatialBN on CPU

### PR DESCRIPTION
**Pull request by xiaomengy**
Summary: Add cuda version of SpatialBNOp also optimize SpatialBN on CPU

Differential Revision: D9512435

### CODE REVIEW

File: caffe2/operators/spatial_batch_norm_op.h [hunk 1]
```diff
@@ -11,69 +16,377 @@ template <class Context>
...
+            C, T(0), running_var->template mutable_data<T>(), &context_);
+      }
+      T* running_mean_data = running_mean->template mutable_data<T>();
+      T* running_var_data = running_var->template mutable_data<T>();
+      if (N == 0) {
+        math::Set<T, Context>(C, T(0), saved_mean_data, &context_);
+        math::Set<T, Context>(C, T(0), saved_rstd_data, &context_);
+        return true;
+      }
+      if (num_batches_ > 1) {
```
Comment (houseroad) [inline caffe2/operators/spatial_batch_norm_op.h]: Add some comments about the computation equation here?

File: caffe2/operators/spatial_batch_norm_op.h [hunk 2]
```diff
@@ -11,69 +16,377 @@ template <class Context>
...
+          N, C, HxW, X_data, alpha_data, beta_data, Y_data, &context_);
+    }
+
     return true;
   }
 
  protected:
-  bool is_test_;
+  template <typename T>
+  void ComputeFusedParam(
```
Comment (houseroad) [inline caffe2/operators/spatial_batch_norm_op.h]: why we call it fusedParam?

File: caffe2/operators/spatial_batch_norm_op.h [hunk 3]
```diff
@@ -11,69 +16,377 @@ template <class Context>
...
+
+  virtual ~SpatialBNGradientOp() = default;
 
   bool RunOnDevice() override {
+    return DispatchHelper<TensorTypes<float>>::call(this, Input(0));
+  }
+
+  template <typename T>
+  bool DoRunWithType() {
+    const auto& X = Input(INPUT);
```
Comment (houseroad) [inline caffe2/operators/spatial_batch_norm_op.h]: Let's just keep the old comments about how to compute the gradients.

**Review by houseroad — APPROVED** [review]
LG, please add some comments about the computation.

File: caffe2/operators/spatial_batch_norm_op.h (hunk 2 above)
Comment (BIT-silence) [reply caffe2/operators/spatial_batch_norm_op.h]: for the Forward pass, the actual formula is Y = (X - mu) / std * scale + bias. In order to reduce the number of computation, we fused it to alpha = scale / std, beta = bias - mu / std * scale. Then the formula is Y = alpha * X + beta. Here we are computing alpha and beta which is the fused param for the forward pass.

**Pull request closed without merging**